src/v/cluster_link/replication/mux_remote_consumer.cc

```diff
@@ -0,0 +1,214 @@
+/*
+ * Copyright 2025 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "cluster_link/replication/mux_remote_consumer.h"
+
+#include "cluster_link/logger.h"
+#include "ssx/future-util.h"
+
+namespace cluster_link::replication {
+
+ss::future<> mux_remote_consumer::start() {
+    co_await _consumer->start();
+    ssx::repeat_until_gate_closed(_gate, [this]() {
+        return fetch_loop().handle_exception([](const std::exception_ptr& e) {
+            if (!ssx::is_shutdown_exception(e)) {
+                vlog(cllog.error, "Fetch loop failed: {}", e);
+            }
+        });
+    });
+}
+
+ss::future<> mux_remote_consumer::stop() noexcept {
+    _cv.broken();
+    auto f = _gate.close();
+    co_await _consumer->stop();
+    co_await std::move(f);
+    co_await ss::max_concurrent_for_each(
+      _partitions, 5, [](auto& p) { return p.second->stop(); });
+}
+
+void mux_remote_consumer::add(const model::topic_partition& tp) {
+    vlog(cllog.trace, "Adding partition {}", tp);
+    auto holder = _gate.hold();
+    vassert(!_partitions.contains(tp), "Duplicate partition added: {}", tp);
+    _partitions.emplace(
+      tp, std::make_unique<partition_data_queue>(_partition_max_buffered));
+    _pending_assignment[tp.topic].insert(tp.partition);
+    _cv.signal();
+}
+
+ss::future<> mux_remote_consumer::remove(const model::topic_partition& tp) {
+    vlog(cllog.trace, "Removing partition {}", tp);
+    auto holder = _gate.hold();
+    auto it = _partitions.find(tp);
+    if (it == _partitions.end()) {
+        co_return;
+    }
+    auto queue = std::move(it->second);
+    _partitions.erase(it);
+    auto pending_it = _pending_assignment.find(tp.topic);
+    if (pending_it != _pending_assignment.end()) {
+        pending_it->second.erase(tp.partition);
+        if (pending_it->second.empty()) {
+            _pending_assignment.erase(pending_it);
+        }
+    }
+    _cv.signal();
+    co_await queue->stop();
+}
+
+void mux_remote_consumer::reset(
+  const model::topic_partition& tp, kafka::offset offset) {
+    auto holder = _gate.hold();
```

Comment:
gate not needed, if you want to prevent this code from being called after the gate is closed `gate.check()` should be sufficient